sql: reorganize EXPLAIN ANALYZE, consolidate schema of EXPLAIN variants

RaduBerinde opened the pull request:
#### sql: reorganize EXPLAIN ANALYZE

This change reorganizes the AST representation for EXPLAIN ANALYZE. We
currently have a separate node for EXPLAIN ANALYZE (DEBUG) but other
EXPLAIN ANALYZE variants use the same tree node as EXPLAIN. The main
reason for this distinction is that the latter is executed very
differently, only supported as a top-level statement.

This change removes the special node for EXPLAIN ANALYZE (DEBUG) and
adds an `ExplainAnalyze` node for all ANALYZE variants. Eventually all
ANALYZE variants will only be supported as top-level statements.

As part of this change, DEBUG is changed from a flag to a mode (which
more closely matches its usage).

Release note: None

#### sql: consolidate schema of EXPLAIN variants

We now have a bunch of EXPLAIN variants that return a single text
column. Consolidate them so they all use the same ResultColumns.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/56353)
<!-- Reviewable:end -->

RaduBerinde commented:
Friendly ping

rytaft reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>  Sorry for the delay!

> Reviewed 16 of 16 files at r1, 12 of 12 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/56353#-:-MLj1GGaEruvYJCh7POj:by6um54)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @mgartner and @RaduBerinde)

___
*[pkg/sql/sem/tree/explain.go, line 27 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/56353#-MLj-1jiDMb3uVu5Sboe:-MLj-1jj0gD3-lBm2rb_:b-2dgvla) ([raw file](https://github.com/cockroachdb/cockroach/blob/fc31b71f948ac9ac2b12138ae38d1cfcb8633116/pkg/sql/sem/tree/explain.go#L27)):*
> ```Go
> 	ExplainOptions
> 
> 	// Statement is the statement being EXPLAINed.
> ```

[nit] would be good to replicate this comment in the `ExplainAnalyze` struct too


<!-- Sent from Reviewable.io -->

RaduBerinde reviewed (commented):
TFTR!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/56353#-:-MLjV6790W5L8fhWsVZZ:b-814g8k)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @mgartner and @rytaft)

___
*[pkg/sql/sem/tree/explain.go, line 27 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/56353#-MLj-1jiDMb3uVu5Sboe:-MLjV5q70CQvrCe6Fhwv:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/fc31b71f948ac9ac2b12138ae38d1cfcb8633116/pkg/sql/sem/tree/explain.go#L27)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

[nit] would be good to replicate this comment in the `ExplainAnalyze` struct too
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

RaduBerinde committed:
sql: reorganize EXPLAIN ANALYZE

This change reorganizes the AST representation for EXPLAIN ANALYZE. We
currently have a separate node for EXPLAIN ANALYZE (DEBUG) but other
EXPLAIN ANALYZE variants use the same tree node as EXPLAIN. The main
reason for this distinction is that the latter is executed very
differently, only supported as a top-level statement.

This change removes the special node for EXPLAIN ANALYZE (DEBUG) and
adds an `ExplainAnalyze` node for all ANALYZE variants. Eventually all
ANALYZE variants will only be supported as top-level statements.

As part of this change, DEBUG is changed from a flag to a mode (which
more closely matches its usage).

Release note: None

RaduBerinde committed:
sql: consolidate schema of EXPLAIN variants

We now have a bunch of EXPLAIN variants that return a single text
column. Consolidate them so they all use the same ResultColumns.

This change also fixes a bug where we were setting up a valuesNode
with the explain columns, which allows in-place modification of those
ResultColumns.

Release note: None

mgartner reviewed (approved):
<img class="emoji" title=":lgtm_strong:" alt=":lgtm_strong:" align="absmiddle" src="https://reviewable.io/lgtm_strong.png" height="20" width="61"/> 

> Reviewed 4 of 16 files at r1, 12 of 12 files at r3, 13 of 13 files at r4.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/56353#-:-MLjcTIBBBzZq0MXSlbi:b-uwkowd)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @RaduBerinde and @rytaft)

___
*[pkg/sql/opt/optbuilder/explain.go, line 93 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/56353#-MLj_L358RnmBCSdT3uy:-MLj_L358RnmBCSdT3uz:bcgvn0x) ([raw file](https://github.com/cockroachdb/cockroach/blob/330d4c679dcadd8b8b7c3e0d9518e41cce65ddf7/pkg/sql/opt/optbuilder/explain.go#L93)):*
> ```Go
> 	b.pushWithFrame()
> 	// We don't allow the statement under Explain to reference outer columns, so we
> 	// pass a "blank" scope rather than inScope.
> ```

Does this mean that `EXPLAIN` can be wrapped in a `SELECT`? Or is there another case when `inScope` wouldn't be blank?


<!-- Sent from Reviewable.io -->

rytaft reviewed (approved):
> Reviewed 13 of 13 files at r4.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/56353)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @RaduBerinde)



<!-- Sent from Reviewable.io -->

RaduBerinde reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/56353)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @mgartner)

___
*[pkg/sql/opt/optbuilder/explain.go, line 93 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/56353#-MLj_L358RnmBCSdT3uy:-MLjfKz-Eh0U2ENLrhQH:b6v8vxo) ([raw file](https://github.com/cockroachdb/cockroach/blob/330d4c679dcadd8b8b7c3e0d9518e41cce65ddf7/pkg/sql/opt/optbuilder/explain.go#L93)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

Does this mean that `EXPLAIN` can be wrapped in a `SELECT`? Or is there another case when `inScope` wouldn't be blank?
</blockquote></details>

yeah, you can do `SELECT * FROM [ EXPLAIN ... ]`. Though we have changed the way the square brackets are handled over time, not 100% this is still strictly needed (can't hurt tho).


<!-- Sent from Reviewable.io -->

RaduBerinde commented:
bors r+

RaduBerinde commented:
bors r+
